consensus/state_processing/src/per_epoch_processing/single_pass.rs

```diff
@@ -0,0 +1,628 @@
+    let total_active_balance = state.get_total_active_balance()?;
+    let churn_limit = state.get_churn_limit(spec)?;
+    let activation_churn_limit = state.get_activation_churn_limit(spec)?;
+    let finalized_checkpoint = state.finalized_checkpoint();
+    let fork_name = state.fork_name_unchecked();
+
+    let state_ctxt = &StateContext {
+        current_epoch,
+        next_epoch,
+        is_in_inactivity_leak,
+        total_active_balance,
+        churn_limit,
+        fork_name,
+    };
+
+    // Contexts that require immutable access to `state`.
+    let slashings_ctxt = &SlashingsContext::new(state, state_ctxt, spec)?;
+    let mut next_epoch_cache = PreEpochCache::new_for_next_epoch(state)?;
+
+    // Split the state into several disjoint mutable borrows.
+    let (
+        validators,
+        balances,
+        previous_epoch_participation,
+        current_epoch_participation,
+        inactivity_scores,
+        progressive_balances,
+        exit_cache,
+        epoch_cache,
+    ) = state.mutable_validator_fields()?;
+
+    let num_validators = validators.len();
+
+    // Take a snapshot of the validators and participation before mutating. This is used for
+    // informational purposes (e.g. by the validator monitor).
+    let summary = ParticipationEpochSummary::new(
+        validators.clone(),
+        previous_epoch_participation.clone(),
+        current_epoch_participation.clone(),
+        previous_epoch,
+        current_epoch,
+    );
+
+    // Compute shared values required for different parts of epoch processing.
+    let rewards_ctxt = &RewardsAndPenaltiesContext::new(progressive_balances, state_ctxt, spec)?;
+    let activation_queue = &epoch_cache
+        .activation_queue()?
+        .get_validators_eligible_for_activation(
+            finalized_checkpoint.epoch,
+            activation_churn_limit as usize,
+        );
+    let effective_balances_ctxt = &EffectiveBalancesContext::new(spec)?;
+
+    // Iterate over the validators and related fields in one pass.
+    let mut validators_iter = validators.iter_mut();
+    let mut balances_iter = balances.iter_mut();
+    let mut inactivity_scores_iter = inactivity_scores.iter_mut();
+
+    // Values computed for the next epoch transition.
+    let mut next_epoch_total_active_balance = 0;
+    let mut next_epoch_activation_queue = ActivationQueue::default();
+
+    for (index, &previous_epoch_participation, &current_epoch_participation) in izip!(
+        0..num_validators,
+        previous_epoch_participation.iter(),
+        current_epoch_participation.iter(),
+    ) {
+        let validator = validators_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
+        let balance = balances_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
+        let inactivity_score = inactivity_scores_iter
+            .next()
+            .ok_or(BeaconStateError::UnknownValidator(index))?;
```

Comment:
Should we keep this syntax or zip all the lists? I would keep as-is to reduce diff latter

Reply:
Agree keeping as-is is good